src/Components/Server/src/Circuits/CircuitSecret.cs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    // Using a type to represent circuit secrets so we can't mistake them
+    // for other kinds of strings/ids.
+    //
+    // Intentionally not overriding ToString here so that this won't accidentally
+    // get logged. It's ok to log the secret at TRACE.
+    internal readonly struct CircuitSecret : IEquatable<CircuitSecret>
+    {
+        public CircuitSecret(string secret)
+        {
+            Secret = secret ?? throw new ArgumentNullException(nameof(secret));
+        }
+
+        public string Secret { get; }
+
+        public bool Equals([AllowNull] CircuitSecret other)
+        {
+            return string.Equals(Secret, other.Secret, StringComparison.Ordinal);
```

Comment:
You can't compare secrets this way. This leaves it open to padding oracle attacks. If you must (I'm not sure why) use `CryptographicOperations.FixedTimeEquals(Span<byte>, Span<byte>)`.

I'm saying this based on the fact that I don't know where this is being used and when, but in principle I would change this just because it's brittle to have the call-site be responsible to do this right (specially if it can cause the actual secret to be leaked).

The only exception where it would be ok to do this (and not entirely) would be if we always check (by unprotecting the payload, which happens in constant time) that any circuit secret sent to us is valid.

I don't want to rely on that, so I would instead do one of the following:
* Use a separate EqualityComparer to do this so that its an opt-in behavior.
* Add the ID to the CircuitSecret and use that for comparisons. That way you don't give up any information about the secret.
* Create a cryptographically secure hash of the secret and use the hash for comparisons. The hash being a one-way function prevents anyone from discovering anything about the secret.

Reply:
I've looked at all the places where equals could be used, the only places that take user input are the DisconnectCircuitMiddleware and ConnectAsync inside the CircuitRegistry, in both cases we do validation of the ID first, so further uses of method are fine.

That said, I think in principle we should change this to use the hash of the circuit for a couple of important benefits:
* We always have a way of going from Secret to ID, which guarantees that a secret and its ID are always paired and there are never mismatches or can be easily detected. (The ID becomes a computed field). This allows us to log based on the ID in all circumstances.
  * No need to enable tracing in logging (which our customers are likely to get wrong and spam their logs with all traces for the server, for example Kestrel will log all the bytes coming in and out). As to do this selectively you would need to use a filter, which is somewhat advanced.
* It allows us to use the ID in all circumstances to do lookups and other operations and allows a safe implementation for Equals and GetHashCode in this type, as even if you are able to figure out the hash for a secret, you don't learn any information about the secret itself (Thanks to hashing being a one way function).

Reply:
Yeah I think you have a point about these being related now. Let's discuss this with security people and if it's ok, I'll use the hash.